File: src/nodes/data/string.rs
use apica_common::bytecodes::ApicaSpecificationBytecode;
use crate::utils::position::Position;

#[derive(Clone)]
pub struct NodeDataString {
    position: Position,
    bytecode: ApicaSpecificationBytecode,
    value: String,
}

impl NodeDataString {
    pub fn init(position: Position, bytecode: ApicaSpecificationBytecode, value: String) -> NodeDataString {
        NodeDataString { position, bytecode, value }
    }
    
    pub fn get_position(&self) -> &Position {
        &self.position
    }
    
    pub fn show(&self, indent: &mut String, end: char) {
        print!("{indent}NodeDataString(data: {:?}, value: {}){end}", self.bytecode, self.value);
    }
    
    pub fn get_bytecode(&self) -> ApicaSpecificationBytecode {
        self.bytecode
    }
    
    pub fn get_value(&self) -> &String {
        &self.value
    }
}

File: src/nodes/data/string.cpp
#include "nodes/data/string.hpp"
#include "core/emitter.hpp"
#include <iostream>

using namespace nodes;

NodeDataString::NodeDataString(const utils::Position &position, common::bytecodes::ApicaSpecificationBytecode spec_bytecode, const std::string &value)
    : NodeData(position, spec_bytecode), value(value) {

}

void NodeDataString::show(std::string &indent, char end) const {
    std::cout << indent << "NodeDataString(spec: " << this->spec_bytecode << ", val: " << this->value << ')' << end;
}

NodeKind NodeDataString::getKind() const {
    return NodeKind::DataString;
}

void NodeDataString::emit(core::Emitter &emitter) const {
    emitter.writeU64(this->spec_bytecode);
    emitter.writeString(this->value);
}

std::string NodeDataString::getSpecificationValue() const {
    return this->value;
}